### DISCUSSION TITLE
Opentelemetry instrumentation not working properly with Ruby 2.6.

### DISCUSSION THREAD

**Discussion by dil-anandp** [open-telemetry/opentelemetry-ruby#1561] [General]
I have used compatible versions of Opentelemetry instrumentation libraries for my Rails app running with image ruby:2.6.6-alpine3.12. 
 
Gemfile

gem 'opentelemetry-sdk', '~> 1.2', '>= 1.2.1'
gem 'opentelemetry-exporter-otlp', '~> 0.24.2'
gem 'opentelemetry-instrumentation-all', '~> 0.33.0'

I followed https://opentelemetry.io/docs/instrumentation/ruby/automatic/.

After completing steps and doing docker-compose it is showing an error given below.

_bundle/gems/opentelemetry-exporter-otlp-0.24.2/lib/opentelemetry/exporter/otlp/exporter.rb:404:in `parse_headers': headers must be a String with comma-separated URL Encoded UTF-8 k=v pairs or a Hash (ArgumentError)_

Is this due to any compatibility issues with Ruby 2.6 and Otel.

**Comment by arielvalentin** [comment]
Please note that the minimum requirements for OTel Ruby are v3.0+ https://opentelemetry.io/docs/instrumentation/ruby/getting-started/#prerequisites

**Reply by dil-anandp** [reply]
Thanks @arielvalentin . So we can't instrument a Ruby app with versions below 3.0 ? Is there any workaround to achieve this?

**Reply by arielvalentin** [reply]
You'll have to downgrade to early versions of the SDK and instrumentations to a version that supported 2.6.

Please keep in mind that versions earlier than 3.0 are no longer receiving security updates, which is why we no longer provide comparability for those versions.

**Reply by dil-anandp** [reply]
I tried with the older versions of SDK and libraries. Sharing below the gems added.

gem 'opentelemetry-sdk', '~> 1.2', '>= 1.2.1'
gem 'opentelemetry-exporter-otlp', '~> 0.24.2'
gem 'opentelemetry-instrumentation-all', '~> 0.33.0'

This is still giving the error mentioned above.

**Reply by arielvalentin** [reply]
Please note that the error has to do with Character encodings of HTTP Headers. Are you sure the `OTEL_EXPORTER_...HEADERS` environment variable is properly formatted and encoded?

**Reply by dil-anandp** [reply]
It is given like this now.
 OTEL_EXPORTER_OTLP_HEADERS: ""
Is this an issue?

**Reply by arielvalentin** [reply]
Yes. It expects at least one key-value pair. It's better to delete it or omit it if it not present. Otherwise add a dummy value.

**Reply by dil-anandp** [reply]
I removed it and now that error is gone. But now I am getting another error

E, [2024-01-10T17:59:53.460539 #23] ERROR -- : OpenTelemetry error: Unable to export 2 spans

**Reply by arielvalentin** [reply]
I am not sure how else to help you other than to ask if your collector or traces service provider is reachable.

**Reply by dil-anandp** [reply]
docker-compose.yml

      OTEL_LOG_LEVEL: 'debug'
      OTEL_TRACES_EXPORTER: 'jaegar'
      OTEL_EXPORTER_OTLP_ENDPOINT: "http://localhost:4318"
      OTEL_EXPORTER_OTLP_TRACES_ENDPOINT: "http://localhost:4318/v1/traces"
      OTEL_EXPORTER_OTLP_TIMEOUT: "10000"
      OTEL_EXPORTER_OTLP_PROTOCOL: "HTTP"
      OTEL_SERVICE_NAME: "minutes-api"
      OTEL_RESOURCE_ATTRIBUTES: "service.namespace=local"

config/initializers/opentelemetry.rb
   
 require 'opentelemetry/sdk'
require 'opentelemetry/exporter/otlp'
require 'opentelemetry/instrumentation/all'

OpenTelemetry::SDK.configure do |c|
  c.use_all
  c.service_name = 'minutes-api'
end

When doing docker-compose services are running without any errors. Jaegar interface is loading with http://localhost:16686/search. But my service minutes-api is not listing there.

When I tried curl http://localhost:4318
curl: (7) Failed to connect to localhost port 4318 after 0 ms: Connection refused.

Jaegar container is listing when doing docker ps.

jaegertracing/all-in-one:latest   "/go/bin/all-in-one-…"   3 days ago     Up 3 days
0.0.0.0:4317-4318->4317-4318/tcp, :::4317-4318->4317-4318/tcp, 0.0.0.0:5778->5778/tcp, :::5778->5778/tcp, 0.0.0.0:9411->9411/tcp, :::9411->9411/tcp, 0.0.0.0:14250->14250/tcp, :::14250->14250/tcp, 0.0.0.0:14268-14269->14268-14269/tcp, :::14268-14269->14268-14269/tcp, 0.0.0.0:16686->16686/tcp, :::16686->16686/tcp, 5775/udp, 0.0.0.0:6831-6832->6831-6832/udp, :::6831-6832->6831-6832/udp   jaeger
d80350a854bf   fb9b574e03c3                      "local-path-provisio…"

**Reply by arielvalentin** [reply]
You're running the jaeger in docker and running your code but then running your app on the host? 

Or is your code also running as a container via docker compose?

Network and DNS rules are different if you are making requests from your host (your pc) from containers are talking to each other through docker.

**Reply by dil-anandp** [reply]
Jaegar and code both are running as containers. Also I verified that they are on same network as well. But still that port is not accessible.

Seeing this error in logs 

jaegar-all-in-one               | {"level":"info","ts":1705035562.0084112,"caller":"grpc@v1.60.0/resolver_wrapper.go:99","msg":"[core][Channel #5] Closing the name resolver","system":"grpc","grpc_log":true}
jaegar-all-in-one               | {"level":"info","ts":1705035562.0086005,"caller":"grpc@v1.60.0/balancer_wrapper.go:178","msg":"[core][Channel #5] ccBalancerWrapper: closing","system":"grpc","grpc_log":true}
jaegar-all-in-one               | {"level":"info","ts":1705035562.0086253,"caller":"grpc@v1.60.0/clientconn.go:532","msg":"[core][Channel #5] Channel Connectivity change to IDLE","system":"grpc","grpc_log":true}
jaegar-all-in-one               | {"level":"info","ts":1705035562.0086365,"caller":"grpc@v1.60.0/clientconn.go:369","msg":"[core][Channel #5] Channel entering idle mode","system":"grpc","grpc_log":true}
jaegar-all-in-one               | {"level":"info","ts":1705035562.0087123,"caller":"grpc/builder.go:131","msg":"Agent collector connection state change","dialTarget":"localhost:14250","status":"IDLE"}
jaegar-all-in-one               | {"level":"info","ts":1705035562.0086765,"caller":"grpc@v1.60.0/clientconn.go:1223","msg":"[core][Channel #5 SubChannel #8] Subchannel Connectivity change to SHUTDOWN","system":"grpc","grpc_log":true}
jaegar-all-in-one               | {"level":"info","ts":1705035562.0088468,"caller":"grpc@v1.60.0/clientconn.go:872","msg":"[core][Channel #5 SubChannel #8] Subchannel deleted","system":"grpc","grpc_log":true}


**jaegar-all-in-one               | {"level":"info","ts":1705034683.0680423,"caller":"grpc@v1.60.0/server.go:983","msg":"[core][Server #4] grpc: Server.Serve failed to create ServerTransport: connection error: desc = \"transport: http2Server.HandleStreams received bogus greeting from client: \\\"GET / HTTP/1.1\\\\r\\\\nHost: lo\\\"\"","system":"grpc","grpc_log":true}**

**Reply by arielvalentin** [reply]
So then the problem is with the jaeger-all-in-one container not accepting connections on 4318 and not a problem with the OTLP Exporter.

Is that correct?

**Reply by dil-anandp** [reply]
Yes. It looks like that. But I tried with adding directly Jaegar service in docker-compose.yml also.

jaegar-all-in-one:
    image: jaegertracing/all-in-one:latest
    ports:
      - "16686:16686"
      - "6831:6831/udp"
      - "6832:6832/udp" 
      - "5778:5778"
      - "4317:4317"
      - "4318:4318"
      - "14250:14250"
      - "14268:14268"
      - "14269:14269"
      - "9411:9411"
    environment:
      COLLECTOR_ZIPKIN_HTTP_PORT: 9411
      COLLECTOR_OTLP_ENABLED: true
    container_name: 'jaegar-all-in-one'

Port is not listening it seems.

When tried to access it via the container shell it gives below result.

**wget http://host.docker.internal:4318
Connecting to host.docker.internal:4318 (192.168.64.1:4318)
wget: can't connect to remote host (192.168.64.1): Operation timed out**


 Is there any issues with the http version mismatch?

**Reply by arielvalentin** [reply]
I think you may want to reach out to the jaeger community for help with this since it's outside of our area of expertise.
